### PR TITLE
Decompile unk_020166C8.s to heap.c

### PR DESCRIPTION
**Pull request by remicalixte**
Welcome to pointer hell.

`heap.h` already contained some of the function declarations so I guess `heap.c` is fitting. All the functions are managing `UnkStruct_020166C8` which seems to be where all the heap information is stored.

### CODE REVIEW

**Review by ProjectRevoTPP — DISMISSED** [review]

File: arm9/src/heap.c [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+THUMB_FUNC u32 FUN_02016834(u32 param0, u32 param1, u32 param2, s32 param3)
+{
+    if (OS_GetProcMode() == OS_PROCMODE_IRQ)
+    {
+        ErrorHandling();
+    }
+
+    u8 *ptr = UNK_021C4D28.unk_ptr5;
+    if (UNK_021C4D28.unk_half4 == ptr[param1])
```
Comment (red031000) [inline arm9/src/heap.c:143]: this feels like it should be a load of GF_ASSERT instead of all the ifs and ErrorHandling etc

File: arm9/src/heap.c [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+
+void FreeToHeap(void *ptr)
+{
+    u8 heap_id = (u8)((u32 *)ptr)[-1];
+
+    if ((u16)heap_id < UNK_021C4D28.unk_half1)
+    {
+        u8 index = UNK_021C4D28.unk_ptr5[heap_id];
+        void *ptr2 = UNK_021C4D28.unk_ptr1[index];
+        if (ptr2 == 0)
```
Comment (red031000) [inline arm9/src/heap.c:301]: GF_ASSERT

File: arm9/src/heap.c [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        OS_RestoreInterrupts(intr_mode);
+        return;
+    }
+
+    ErrorHandling();
+}
+
+void FUN_02016A8C(u32 param0, void *param1)
+{
+    if (OS_GetProcMode() == OS_PROCMODE_IRQ)
```
Comment (red031000) [inline arm9/src/heap.c:326]: GF_ASSERT

File: arm9/src/heap.c [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    {
+        ErrorHandling();
+    }
+
+    if (param0 < UNK_021C4D28.unk_half1)
+    {
+        u8 index = UNK_021C4D28.unk_ptr5[param0];
+        void *ptr = UNK_021C4D28.unk_ptr1[index];
+        if (ptr == 0)
```
Comment (red031000) [inline arm9/src/heap.c:335]: GF_ASSERT (for the three below as well)

File: arm9/src/heap.c [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        FUN_020AE82C(param0, UNK_021C4D28.unk_ptr1[index], param2);
+        return;
+    }
+
+    ErrorHandling();
+}
+
+THUMB_FUNC void FUN_02016B44(void *ptr, u32 param1)
+{
+    if (OS_GetProcMode() == OS_PROCMODE_IRQ)
```
Comment (red031000) [inline arm9/src/heap.c:386]: GF_ASSERT

File: include/heap.h [hunk 6]
```diff
@@ -1,9 +1,37 @@
 #ifndef POKEDIAMOND_HEAP_H
 #define POKEDIAMOND_HEAP_H
 
-void * AllocFromHeap(u32 heap_id, u32 size);
-void * AllocFromHeapAtEnd(u32 heap_id, u32 size);
-void FreeToHeap(void * ptr);
+#include "global.h"
+
+struct UnkStruct_020166C8
```
Comment (red031000) [inline include/heap.h:6]: this isn't normally how struct members are named, unk00 for item at 0x00 etc etc

**Review by red031000 — CHANGES_REQUESTED** [review]

File: arm9/src/heap.c (hunk 1 above)
Comment (remicalixte) [reply arm9/src/heap.c:143]: Ah nice, didn't about this macro

**Review by red031000 — APPROVED** [review]

**Pull request merged**